subprojects/docs/src/docs/userguide/developing-plugins/kotlin_gradle_plugins.adoc

```diff
@@ -0,0 +1,334 @@
+dependencies {
+    compileOnly("org.jetbrains.kotlin:kotlin-stdlib:1.4.31")
+}
+----
+
+For similar reasons, any transitive dependency of your plugin must use a `kotlin-stdlib` version compatible with the Gradle one.
+While you can ensure that at compile time for your code, it's often more complicated for dependencies, so test your plugins and/or double-check your transitive dependencies.
+
+Note: The `kotlin-stdlib` version is independent of your Kotlin Gradle Plugin version. You can use a newer Kotlin Gradle Plugin version with an older `kotlin-stdlib` using the https://kotlinlang.org/docs/compiler-reference.html#api-version-version[`apiVersion`] compiler argument.
+
+TIP: Because it is provided by Gradle at runtime, `kotlin-stdlib` must be a `compileOnly` dependency.
+
+[[sec:kotlin_reflect]]
+=== kotlin-reflect
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly("org.jetbrains.kotlin:kotlin-reflect:$version")
+}
+----
+
+Not all plugins need `kotlin-reflect` but if yours does, the situation is the same as for `kotlin-stdlib`. `kotlin-reflect` needs to be compatible with the https://docs.gradle.org/current/userguide/compatibility.html#kotlin[target Gradle version]
+
+TIP: Because it is provided by Gradle at runtime, `gradle-api` must be a `compileOnly` dependency.
+
+[[sec:others]]
+=== others (`implementation` or `api`)
+
+You can add other non-compileOnly dependencies like in other JVM projects:
+
+[source,kotlin]
+----
+dependencies {
+    // Kotlin dependencies
+    // ⚠️ make sure they do not depend on an incompatible version of kotlin-stdlib
+    implementation("com.squareup:kotlinpoet:1.12.0")
+
+    // Java dependencies
+    implementation("asm:asm:3.3.1")
+}
+----
+
+Gradle does not provide these dependencies at runtime and should use the `implementation` configuration (or `api` if part of your plugin API). Because Gradle may load different plugins with conflicting dependencies in the same https://dev.to/autonomousapps/build-compile-run-a-crash-course-in-classpaths-f4g[classloader], your plugin users may experience crashes if symbols are incompatible or missing. To mitigate those, you can relocate your dependencies.
+
+[[sec:compiler_options]]
+== Different compiler configurations
+
+You can choose different options to either improve the developer experience of writing plugins and/or improve compatibility with Gradle and Java in general.
+
+[[sec:sam_with_receiver]]
+=== `sam-with-receiver` compiler plugin
+
+Gradle comes with the https://docs.gradle.org/current/javadoc/org/gradle/api/HasImplicitReceiver.html[@HasImplicitReceiver] annotation. When used in conjunction with Kotlin https://kotlinlang.org/docs/sam-with-receiver-plugin.html[sam-with-receiver compiler plugin], it can turn the parameter of https://kotlinlang.org/docs/fun-interfaces.html[Single Abstract Method (SAM)] interfaces like https://docs.gradle.org/current/javadoc/org/gradle/api/Action.html[Action] into implicit receiver making your code look more like a DSL:
```

Comment:
```suggestion
Gradle comes with the https://docs.gradle.org/current/javadoc/org/gradle/api/HasImplicitReceiver.html[@HasImplicitReceiver] annotation.
When used in conjunction with Kotlin https://kotlinlang.org/docs/sam-with-receiver-plugin.html[sam-with-receiver compiler plugin], it can turn the parameter of https://kotlinlang.org/docs/fun-interfaces.html[Single Abstract Method (SAM)] interfaces like https://docs.gradle.org/current/javadoc/org/gradle/api/Action.html[Action] into implicit receiver making your code look more like a DSL:
```